Refactor: Implementing (some of) AppController2

chynu opened the pull request:
Implemented refactored functions in `app_controller_2.js` as well as referenced functions in `onTab()` in `app_controller_2.js`.

For future: Deprecate Block Exporter tab and break apart Workspace Factory into Toolbox Editor and Workspace Editor. Left stubs/minimal functionality for block exporter/workspace factory for now, but will be deleted soonin upcoming PR's after the proper views are added.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/google/blockly-devtools/96)
<!-- Reviewable:end -->

chynu committed:
Started implementing functions in AppController2

chynu committed:
Implemented tab switching in AppController2

chynu committed:
Merge branch 'develop' into rf_appcontroller

chynu committed:
Implemented createSampleApplication()

chynu committed:
Nits

chynu committed:
Merge branch 'develop' into rf_appcontroller

chynu committed:
Minor changes

svouse commented:
Review status: 0 of 11 files reviewed at latest revision, 1 unresolved discussion.

---

*[res/devtools_toolboxes.js, line 55 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi43EMq--2T6KVEG1z:-Koi43EMq--2T6KVEG2-:bua3700) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/res/devtools_toolboxes.js#L55)):*
> ```JavaScript
>  *     custom Toolboxes and WorkspaceContents.
>  */
> DevToolsToolboxes.toolboxEditor = function(blockLibraryMap) {
> ```

Would it be useful at all to have a function for creating the xml string in BlockLibrarySet, then just pass the result to this function and remove the need for this loop being here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/96)*
<!-- Sent from Reviewable.io -->

rachel-fenichel commented:
Reviewed 7 of 11 files at r1.
Review status: 7 of 11 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/controller/app_controller_2.js, line 183 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi7dvvZ67z2oOSkymS:-Koi7dvvZ67z2oOSkymT:bixdas) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/src/controller/app_controller_2.js#L183)):*
> ```JavaScript
>     if (this.view.selectedTab == AppController2.BLOCK_FACTORY) {
>       // Hide other tabs.
>       FactoryUtils.hide('blockLibraryExporter');
> ```

Most of the code here is of the form
hide (thing)
hide (other thing)
hide (third thing)
show (last thing)

Instead you could have a factory utils function that takes in an array of things to hide and a single thing to show.  Then it would be
functionName ([thing, other thing, third thing], last thing)

An alternative is to pass in only the thing you want to show, and have it call hide on everything else:
factoryUtils.showOnly(last thing)

---

*[src/controller/workspace_controller.js, line 314 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi6n1_-I0MI7ez8pk2:-Koi6n1_-I0MI7ez8pk3:bepdpv) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/src/controller/workspace_controller.js#L314)):*
> ```JavaScript
>   /*
>    * Updates the block library category in the toolbox workspace toolbox.
>    * @param {!Element} categoryXml XML for the block library category.
> ```

JSdoc should match functions.  You have no arguments now but still have doc for them.

---

*[src/controller/workspace_controller.js, line 323 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi7MoDPFYQVCoRDvRY:-Koi7MoDPFYQVCoRDvRZ:b95ihwd) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/src/controller/workspace_controller.js#L323)):*
> ```JavaScript
>      *       (Also moved into: toolbox_controller.js)
>      */
>     const libraryXmls = {};
> ```

This code to build the libraryXmls should be shared between workspaceController and toolboxController instead of duplicated.

---

*[src/controller/workspace_controller.js, line 326 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi7AR-1coTMZZ-ZTFl:-Koi7AR-1coTMZZ-ZTFm:bi74f1w) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/src/controller/workspace_controller.js#L326)):*
> ```JavaScript
>     const libraryNames = this.project.getLibraryNames();
> 
>     for (const i in libraryNames) {
> ```

If `libraryNames` is an array I would expect `i` to be elements in the array, not a counter.  Please check that this function works.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/96)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
Review status: 7 of 11 files reviewed at latest revision, 13 unresolved discussions.

---

*[res/devtools_toolboxes.js, line 40 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi3HmKVgMVtPML8JX-:-Koi3HmLrcsBARkX3XQ1:bihcwul) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/res/devtools_toolboxes.js#L40)):*
> ```JavaScript
> DevToolsToolboxes.createCategoryElement_ = function(libraryName, libraryXml) {
>   // TODO: Create unique document IDs for each block library. Currently all
>   //       block libraries have the same ID, "blockLibrary".
> ```

Just remove the category id. We shouldn't depend upon it, until we are giving users an edit UI for the field.

---

*[src/controller/app_controller_2.js, line 160 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi4esKVLKoD48Y5goa:-Koi4esKVLKoD48Y5gob:b225vdp) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/src/controller/app_controller_2.js#L160)):*
> ```JavaScript
>   onTab() {
>     /*
>      * TODO: Move in from app_controller.js
> ```

Remove TODO

---

*[src/controller/app_controller_2.js, line 271 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi5FeVfSuVSOQLyxi6:-Koi5FeVfSuVSOQLyxi7:b-sx2s3f) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/src/controller/app_controller_2.js#L271)):*
> ```JavaScript
>   initializeBlocklyStorage() {
>     // REFACTORED: Moved in from app_controller.js
>     BlocklyStorage.HTTPREQUEST_ERROR =
> ```

Unless there is an obvious reason not to, let's make these constants  (or static getters) on BlocklyStorage, instead of assigning them here?

---

*[src/controller/toolbox_controller.js, line 60 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi67Jssm16G5XD1-co:-Koi67Jssm16G5XD1-cp:b-cuhb0z) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/src/controller/toolbox_controller.js#L60)):*
> ```JavaScript
>      * @type {boolean}
>      */
>     this.keyEventsEnabled = true;
> ```

I don't see a remove block that matches this, or other usage in this file. I'm not clear where this came from, or how it is/was used.

---

*[src/controller/toolbox_controller.js, line 562 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi6SSiMkylGceN_LJ2:-Koi6SSiMkylGceN_LJ3:b3oikuv) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/src/controller/toolbox_controller.js#L562)):*
> ```JavaScript
>      */
>     const libraryXmls = {};
>     const libraryNames = this.project.getLibraryNames();
> ```

Somehow note why you need the names separate from the map. Either add a comment noting that they are in user order, or rename libraryNames as either libraryNamesOrdered or libraryNamesArray.

---

*[src/controller/workspace_controller.js, line 313 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi88HhSGxdLvp9vjZ1:-Koi88HhSGxdLvp9vjZ2:b-72t6l0) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/src/controller/workspace_controller.js#L313)):*
> ```JavaScript
> 
>   /*
>    * Updates the block library category in the toolbox workspace toolbox.
> ```

No longer "toolbox workspace toolbox"

---

*[src/controller/workspace_controller.js, line 318 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi7Pb_JQfub7BWYjXV:-Koi7Pb_JQfub7BWYjXW:borw0xv) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/src/controller/workspace_controller.js#L318)):*
> ```JavaScript
>    *     library.
>    */
>   setBlockLibCategory() {
> ```

This doesn't seem like a `set` method anymore (no input params).  Maybe just `update`?

---

*[src/controller/workspace_controller.js, line 323 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi7MoDPFYQVCoRDvRY:-Koi8Idc5UMPKUBnygMX:b32v57) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/src/controller/workspace_controller.js#L323)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

This code to build the libraryXmls should be shared between workspaceController and toolboxController instead of duplicated.
</blockquote></details>

+1

---

*[src/controller/workspace_controller.js, line 326 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi7AR-1coTMZZ-ZTFl:-Koi8M55oiLzwpHsW5wY:b5ooiut) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/src/controller/workspace_controller.js#L326)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

If `libraryNames` is an array I would expect `i` to be elements in the array, not a counter.  Please check that this function works.
</blockquote></details>

It is correctly an array index for JavaScripts's variant of for..in: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in

`Array.foreach()` can be less confusing, in this regard.

---

*[src/model/block_library.js, line 230 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi81hhfqdV19Bl4FNS:-Koi81hhfqdV19Bl4FNT:b9sikyb) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/src/model/block_library.js#L230)):*
> ```JavaScript
>    * @return {!Element} XML representation of the block library category.
>    */
>   getCategoryXml() {
> ```

Move this to ToolboxEditor

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/96)*
<!-- Sent from Reviewable.io -->

rachel-fenichel commented:
Review status: 7 of 11 files reviewed at latest revision, 13 unresolved discussions.

---

*[src/controller/workspace_controller.js, line 326 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi7AR-1coTMZZ-ZTFl:-Koi9Nq6B2nNCaQM77rO:bhrjvx) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/src/controller/workspace_controller.js#L326)):*
Okay.  It should be `for...of` or `forEach` since the index isn't being used.

Also note that 

>Array indexes are just enumerable properties with integer names and are otherwise identical to general Object properties. There is no guarantee that for...in will return the indexes in any particular order.  The for...in loop statement will return all enumerable properties, including those with non–integer names and those that are inherited.

>Because the order of iteration is implementation-dependent, iterating over an array may not visit elements in a consistent order. Therefore it is better to use a for loop with a numeric index (or Array.prototype.forEach() or the for...of loop) when iterating over arrays where the order of access is important.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/96)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
Review status: 7 of 11 files reviewed at latest revision, 13 unresolved discussions.

---

*[src/controller/workspace_controller.js, line 326 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi7AR-1coTMZZ-ZTFl:-KoiimKGkM9JU2XKb5pa:b-fnm29x) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/src/controller/workspace_controller.js#L326)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Okay.  It should be `for...of` or `forEach` since the index isn't being used.

Also note that 

>Array indexes are just enumerable properties with integer names and are otherwise identical to general Object properties. There is no guarantee that for...in will return the indexes in any particular order.  The for...in loop statement will return all enumerable properties, including those with non–integer names and those that are inherited.

>Because the order of iteration is implementation-dependent, iterating over an array may not visit elements in a consistent order. Therefore it is better to use a for loop with a numeric index (or Array.prototype.forEach() or the for...of loop) when iterating over arrays where the order of access is important.
</blockquote></details>

Cool.  New ES6 syntax I didn't know.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/96)*
<!-- Sent from Reviewable.io -->

chynu committed:
Fixes for #96

chynu commented:
Review status: 3 of 11 files reviewed at latest revision, 13 unresolved discussions.

---

*[res/devtools_toolboxes.js, line 40 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi3HmKVgMVtPML8JX-:-Koi9sybeO3TLKXhla-_:bblz4u9) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/res/devtools_toolboxes.js#L40)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Just remove the category id. We shouldn't depend upon it, until we are giving users an edit UI for the field.
</blockquote></details>

Makes sense. Removed !

---

*[res/devtools_toolboxes.js, line 55 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi43EMq--2T6KVEG1z:-KoiA9GINDDzDBLomtrA:brjufkb) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/res/devtools_toolboxes.js#L55)):*
<details><summary><i>Previously, svouse wrote…</i></summary><blockquote>

Would it be useful at all to have a function for creating the xml string in BlockLibrarySet, then just pass the result to this function and remove the need for this loop being here?
</blockquote></details>

Hm, that could also work. I suppose we could leave this as is for now and once it is merged into develop you can add changes if you create that functionality on the model level.

---

*[src/controller/app_controller_2.js, line 160 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi4esKVLKoD48Y5goa:-KoiAKLcQaT8J4i_N9Y5:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/src/controller/app_controller_2.js#L160)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Remove TODO
</blockquote></details>

Done.

---

*[src/controller/app_controller_2.js, line 183 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi7dvvZ67z2oOSkymS:-KoiS6brQW03gRbUv-L0:b-xhhmoe) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/src/controller/app_controller_2.js#L183)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Most of the code here is of the form
hide (thing)
hide (other thing)
hide (third thing)
show (last thing)

Instead you could have a factory utils function that takes in an array of things to hide and a single thing to show.  Then it would be
functionName ([thing, other thing, third thing], last thing)

An alternative is to pass in only the thing you want to show, and have it call hide on everything else:
factoryUtils.showOnly(last thing)
</blockquote></details>

Done. Created a helper function, `showTab_()` right below `onTab()`.

---

*[src/controller/app_controller_2.js, line 271 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi5FeVfSuVSOQLyxi6:-KoiLnmLxDF_6AdGGKCT:b3d53wc) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/src/controller/app_controller_2.js#L271)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Unless there is an obvious reason not to, let's make these constants  (or static getters) on BlocklyStorage, instead of assigning them here?
</blockquote></details>

I couldn't find BlocklyStorage in this repo (it is something in AppEngine?). I think we will discuss this later today (as per a brief discussion earlier today), so I will leave this comment unaddressed for now.

---

*[src/controller/toolbox_controller.js, line 60 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi67Jssm16G5XD1-co:-KoiM3cYA_hiMbbRX5PI:b7xhe0s) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/src/controller/toolbox_controller.js#L60)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

I don't see a remove block that matches this, or other usage in this file. I'm not clear where this came from, or how it is/was used.
</blockquote></details>

I think it will be used in other functions that haven't yet been implemented, but this field will likely be removed after all functions are filled in. It doesn't seem entirely necessary to me, but I included it for now just in case.

---

*[src/controller/toolbox_controller.js, line 562 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi6SSiMkylGceN_LJ2:-KoiNKtkmmKBTZolYtt9:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/src/controller/toolbox_controller.js#L562)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Somehow note why you need the names separate from the map. Either add a comment noting that they are in user order, or rename libraryNames as either libraryNamesOrdered or libraryNamesArray.
</blockquote></details>

Done.

---

*[src/controller/workspace_controller.js, line 313 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi88HhSGxdLvp9vjZ1:-KoiNi7zBkH1wI-JwRzX:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/src/controller/workspace_controller.js#L313)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

No longer "toolbox workspace toolbox"
</blockquote></details>

Done.

---

*[src/controller/workspace_controller.js, line 314 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi6n1_-I0MI7ez8pk2:-KoiNqDIrgRJSiTWrgAp:b-a4wcvi) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/src/controller/workspace_controller.js#L314)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

JSdoc should match functions.  You have no arguments now but still have doc for them.
</blockquote></details>

Oops. Done.

---

*[src/controller/workspace_controller.js, line 318 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi7Pb_JQfub7BWYjXV:-KoiNsJ1pxOnTnr7NzpQ:biavqw8) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/src/controller/workspace_controller.js#L318)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

This doesn't seem like a `set` method anymore (no input params).  Maybe just `update`?
</blockquote></details>

Renamed to `updateBlockLibCategory()`.

---

*[src/controller/workspace_controller.js, line 323 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi7MoDPFYQVCoRDvRY:-KoiNvKS8GrCvPnGDCYa:b6ljgtr) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/src/controller/workspace_controller.js#L323)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

+1
</blockquote></details>

I moved the function to the `EditorController` level.

---

*[src/controller/workspace_controller.js, line 326 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi7AR-1coTMZZ-ZTFl:-KoiOGdJ7VG0BOzHG85Q:brd86q) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/src/controller/workspace_controller.js#L326)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Cool.  New ES6 syntax I didn't know.
</blockquote></details>

I have changed it to use `Array.forEach()`. Thank you !

I'm assuming it's okay to use `Array.forEach()`, `for ... of` interchangeably (i.e. that there are no significant advantages/stylistic preferences towards one or the other) for Blockly.

---

*[src/model/block_library.js, line 230 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi81hhfqdV19Bl4FNS:-KoiSjmTPbq-1fkRLSeb:b-t4hn4o) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/src/model/block_library.js#L230)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Move this to ToolboxEditor
</blockquote></details>

Moved to `EditorController`, since both the toolbox and workspace editor will be using the same exact code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/96)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
Review status: 3 of 11 files reviewed at latest revision, 14 unresolved discussions.

---

*[res/devtools_toolboxes.js, line 55 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi43EMq--2T6KVEG1z:-KoinY1LoiDN1_h18Eox:b-lg7x7) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/res/devtools_toolboxes.js#L55)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

Hm, that could also work. I suppose we could leave this as is for now and once it is merged into develop you can add changes if you create that functionality on the model level.
</blockquote></details>

IMO, this feels like the right place for this function.

---

*[res/devtools_toolboxes.js, line 55 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/96#-KoinCdi6U_HSUjPGygu:-KoinCdi6U_HSUjPGygv:b5040sb) ([raw file](https://github.com/google/blockly-devtools/blob/06d2c885981fdf8ba774727d98bc332b35cceb6b/res/devtools_toolboxes.js#L55)):*
> ```JavaScript
> DevToolsToolboxes.toolboxEditor = function(blockLibraryMap) {
>   const blockLibraryXmls = '';
>   for (const blockLibName in blockLibraryMap) {
> ```

Hmmm... This loop is building an ordered string out of the arbitrary order of the map keys.

---

*[src/controller/app_controller_2.js, line 271 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi5FeVfSuVSOQLyxi6:-KoiqQYLJk2Y_kOOHcTd:b-j6mnl5) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/src/controller/app_controller_2.js#L271)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

I couldn't find BlocklyStorage in this repo (it is something in AppEngine?). I think we will discuss this later today (as per a brief discussion earlier today), so I will leave this comment unaddressed for now.
</blockquote></details>

Per offline discussion, ignore for now.  I would even recommend deleting this method.

---

*[src/controller/app_controller_2.js, line 296 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koio16WDsczvwjqaojs:-Koio16WDsczvwjqaojt:bwwlulp) ([raw file](https://github.com/google/blockly-devtools/blob/06d2c885981fdf8ba774727d98bc332b35cceb6b/src/controller/app_controller_2.js#L296)):*
> ```JavaScript
>       return;
>     }
>     const data = new Blob([injectFileContents], {type: 'text/javascript'});
> ```

What is the reason for moving the Blob out of the utility function? I don't see it used in other ways.

---

*[src/controller/editor_controller.js, line 69 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/96#-KoivUn51YE8qoBFu0FQ:-KoivUn51YE8qoBFu0FR:bz0ujt) ([raw file](https://github.com/google/blockly-devtools/blob/06d2c885981fdf8ba774727d98bc332b35cceb6b/src/controller/editor_controller.js#L69)):*
> ```JavaScript
>      * @type {!Object.<string, Blockly.Workspace}
>      */
>     this.hiddenWorkspaces = {
> ```

Move these to their respective editor controller.

---

*[src/controller/editor_controller.js, line 153 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/96#-KoiuwDVkeKt1eyId2OG:-KoiuwDVkeKt1eyId2OH:bn1dzsy) ([raw file](https://github.com/google/blockly-devtools/blob/06d2c885981fdf8ba774727d98bc332b35cceb6b/src/controller/editor_controller.js#L153)):*
> ```JavaScript
>    * toolboxes.
>    */
>   updateBlockLibCategory() {
> ```

This method has nothing to do with EditorController, and is independent enough to be a utility function.

---

*[src/controller/editor_controller.js, line 156 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koiu3IQzlqMJKDz2J3K:-Koiu3IQzlqMJKDz2J3L:bkhuxe8) ([raw file](https://github.com/google/blockly-devtools/blob/06d2c885981fdf8ba774727d98bc332b35cceb6b/src/controller/editor_controller.js#L156)):*
> ```JavaScript
>     /*
>      * REFACTORED: Moved in from wfactory_controller.js
>      *       (Also moved into: workspace_controller.js)
> ```

No longer moved into worksapce_controller.

---

*[src/controller/editor_controller.js, line 169 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/96#-KoiuTXlZ32UOEFVAc8F:-KoiuTXlZ32UOEFVAc8G:b6ru2ix) ([raw file](https://github.com/google/blockly-devtools/blob/06d2c885981fdf8ba774727d98bc332b35cceb6b/src/controller/editor_controller.js#L169)):*
> ```JavaScript
> 
>     // Update editor toolboxes with new libraries.
>     const newToolboxXml = DevToolsToolboxes.toolboxEditor(libraryXmls);
> ```

Return `newToolboxXml`. Allow each of `toolboxController` and `workspaceController` to update themselves.

---

*[src/controller/editor_controller.js, line 181 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/96#-KoivFp8mHtRzbmWonvs:-KoivFp8mHtRzbmWonvt:b5xzps6) ([raw file](https://github.com/google/blockly-devtools/blob/06d2c885981fdf8ba774727d98bc332b35cceb6b/src/controller/editor_controller.js#L181)):*
> ```JavaScript
>    * @return {!Element} XML representation of the block library category.
>    */
>   getCategoryXml(library) {
> ```

Also a utility method.

---

*[src/controller/workspace_controller.js, line 323 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi7MoDPFYQVCoRDvRY:-KoimUcrnmL0zVkRA0Xt:b-vddag0) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/src/controller/workspace_controller.js#L323)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

I moved the function to the `EditorController` level.
</blockquote></details>

Per other comments, make it a utility function.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/96)*
<!-- Sent from Reviewable.io -->

chynu committed:
Merge branch 'develop' into rf_appcontroller

chynu committed:
Fixes for #96

chynu committed:
Unified hidden workspaces to AppController level

chynu commented:
Review status: 3 of 11 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[res/devtools_toolboxes.js, line 55 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/96#-KoinCdi6U_HSUjPGygu:-Kowaov7J0B87P6fwPDK:b-79xz0v) ([raw file](https://github.com/google/blockly-devtools/blob/06d2c885981fdf8ba774727d98bc332b35cceb6b/res/devtools_toolboxes.js#L55)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Hmmm... This loop is building an ordered string out of the arbitrary order of the map keys.
</blockquote></details>

Would it be a better idea to make it into an array?

---

*[src/controller/app_controller_2.js, line 271 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi5FeVfSuVSOQLyxi6:-KoizsKAQU1SgH1dnWHU:b-dg2b2t) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/src/controller/app_controller_2.js#L271)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Per offline discussion, ignore for now.  I would even recommend deleting this method.
</blockquote></details>

I added a comment for now.

---

*[src/controller/app_controller_2.js, line 296 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koio16WDsczvwjqaojs:-Koj-BLG5yEGaYbK4UhB:b8uxnti) ([raw file](https://github.com/google/blockly-devtools/blob/06d2c885981fdf8ba774727d98bc332b35cceb6b/src/controller/app_controller_2.js#L296)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

What is the reason for moving the Blob out of the utility function? I don't see it used in other ways.
</blockquote></details>

I thought that we might use it differently in the future, though I am not entirely sure how Blob works in the context of downloading multiple files (or if Blob is used in such cases).

---

*[src/controller/editor_controller.js, line 69 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/96#-KoivUn51YE8qoBFu0FQ:-Kow_vxfcDXmHI5MMyWp:b-8wk2x7) ([raw file](https://github.com/google/blockly-devtools/blob/06d2c885981fdf8ba774727d98bc332b35cceb6b/src/controller/editor_controller.js#L69)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Move these to their respective editor controller.
</blockquote></details>

I unified all hidden workspaces into one `Blockly.Workspace` instantiated at the `AppController` level. There are field references to the same workspace in `ToolboxController`, `WorkspaceController`, and `BlockEditorController`. I did this because the hidden workspaces are only used as a temporary means to create Blockly Blocks or some other Blockly object and is always cleared before/after using it within the function. Otherwise I would have had to create a Blockly Workspace for each sub-controller class.

---

*[src/controller/editor_controller.js, line 153 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/96#-KoiuwDVkeKt1eyId2OG:-KojDOBk21Rvb-anUuOR:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/06d2c885981fdf8ba774727d98bc332b35cceb6b/src/controller/editor_controller.js#L153)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

This method has nothing to do with EditorController, and is independent enough to be a utility function.
</blockquote></details>

Done.

---

*[src/controller/editor_controller.js, line 156 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koiu3IQzlqMJKDz2J3K:-KojDXJ1Xw5wtT0kwRkg:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/06d2c885981fdf8ba774727d98bc332b35cceb6b/src/controller/editor_controller.js#L156)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

No longer moved into worksapce_controller.
</blockquote></details>

Done.

---

*[src/controller/editor_controller.js, line 169 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/96#-KoiuTXlZ32UOEFVAc8F:-KojFbVGP-oq3ljJLmQc:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/06d2c885981fdf8ba774727d98bc332b35cceb6b/src/controller/editor_controller.js#L169)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Return `newToolboxXml`. Allow each of `toolboxController` and `workspaceController` to update themselves.
</blockquote></details>

Done.

---

*[src/controller/editor_controller.js, line 181 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/96#-KoivFp8mHtRzbmWonvs:-KojFastxFGMU22H9Pdu:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/06d2c885981fdf8ba774727d98bc332b35cceb6b/src/controller/editor_controller.js#L181)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Also a utility method.
</blockquote></details>

Done.

---

*[src/controller/workspace_controller.js, line 323 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi7MoDPFYQVCoRDvRY:-KowafISyfODTrnK2dNJ:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/src/controller/workspace_controller.js#L323)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Per other comments, make it a utility function.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/96)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
<img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>  If you want to deal with the map sorting issues later, that's fine.

---

Reviewed 4 of 11 files at r1, 1 of 7 files at r2, 7 of 8 files at r3.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[res/devtools_toolboxes.js, line 55 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/96#-KoinCdi6U_HSUjPGygu:-KowpNUGoT6n_RAh-6s3:bxazcvn) ([raw file](https://github.com/google/blockly-devtools/blob/06d2c885981fdf8ba774727d98bc332b35cceb6b/res/devtools_toolboxes.js#L55)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

Would it be a better idea to make it into an array?
</blockquote></details>

Maybe just sort the map's keys into alphabetical order?

---

*[src/factory_utils.js, line 704 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/96#-KowqNvqkxRzMInMt2LM:-KowqNvqkxRzMInMt2LN:b-4pect4) ([raw file](https://github.com/google/blockly-devtools/blob/599e2c4d9cf48461a09b15f2d2becf94d6d5f38f/src/factory_utils.js#L704)):*
> ```JavaScript
>  * @param {string} filename Name of file to download.
>  */
> FactoryUtils.createAndDownloadFile = function(data, filename) {
> ```

Revert to the old form, where the Blob is handled here, except pass in the full MIME type ('text/plain', 'application/javascript', etc.) for the final parameter as `mimeType`.

---

*[src/controller/app_controller_2.js, line 296 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koio16WDsczvwjqaojs:-Kowq5ags2c8hJ0bnZuy:bix8tic) ([raw file](https://github.com/google/blockly-devtools/blob/06d2c885981fdf8ba774727d98bc332b35cceb6b/src/controller/app_controller_2.js#L296)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

I thought that we might use it differently in the future, though I am not entirely sure how Blob works in the context of downloading multiple files (or if Blob is used in such cases).
</blockquote></details>

I don't think there is any value to exposing the blob. Revert to the old form, 

---

*[src/controller/editor_controller.js, line 69 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/96#-KoivUn51YE8qoBFu0FQ:-Kowr-uWqveUeqm_KfXB:b-ovwaw7) ([raw file](https://github.com/google/blockly-devtools/blob/06d2c885981fdf8ba774727d98bc332b35cceb6b/src/controller/editor_controller.js#L69)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

I unified all hidden workspaces into one `Blockly.Workspace` instantiated at the `AppController` level. There are field references to the same workspace in `ToolboxController`, `WorkspaceController`, and `BlockEditorController`. I did this because the hidden workspaces are only used as a temporary means to create Blockly Blocks or some other Blockly object and is always cleared before/after using it within the function. Otherwise I would have had to create a Blockly Workspace for each sub-controller class.
</blockquote></details>

Good idea.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/96#-:-Kowt0R1_ZxvQhW8_Kpi:b-gofeo2)*
<!-- Sent from Reviewable.io -->

chynu committed:
Fixes for #96

chynu commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[res/devtools_toolboxes.js, line 55 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/96#-KoinCdi6U_HSUjPGygu:-Kox577efW_MhHh31DJb:b-ukp1et) ([raw file](https://github.com/google/blockly-devtools/blob/06d2c885981fdf8ba774727d98bc332b35cceb6b/res/devtools_toolboxes.js#L55)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Maybe just sort the map's keys into alphabetical order?
</blockquote></details>

I haven't implemented it in this PR because I think it is best if the library names (and other resource names) are alphabetized at the model level (in `BlockLibrarySet`, for example).

---

*[src/factory_utils.js, line 704 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/96#-KowqNvqkxRzMInMt2LM:-Kox68FIO_p_YPfHK0mi:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/599e2c4d9cf48461a09b15f2d2becf94d6d5f38f/src/factory_utils.js#L704)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Revert to the old form, where the Blob is handled here, except pass in the full MIME type ('text/plain', 'application/javascript', etc.) for the final parameter as `mimeType`.
</blockquote></details>

Done.

---

*[src/controller/app_controller_2.js, line 296 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koio16WDsczvwjqaojs:-Kox6ALrdhiveoHMDg_b:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/06d2c885981fdf8ba774727d98bc332b35cceb6b/src/controller/app_controller_2.js#L296)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

I don't think there is any value to exposing the blob. Revert to the old form, 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/96)*
<!-- Sent from Reviewable.io -->

chynu committed:
Merge branch 'develop' into rf_appcontroller

rachel-fenichel commented:
Reviewed 1 of 11 files at r1, 5 of 7 files at r2, 3 of 8 files at r3, 2 of 7 files at r4.
Review status: 9 of 14 files reviewed at latest revision, 9 unresolved discussions.

---

*[src/controller/app_controller_2.js, line 183 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi7dvvZ67z2oOSkymS:-KoiuZj_LlOYzLzadOrl:bp6uxwa) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/src/controller/app_controller_2.js#L183)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

Done. Created a helper function, `showTab_()` right below `onTab()`.
</blockquote></details>

Sweet.  This got a lot smaller.

What happens if multiple tabs are told to show at the same time?

---

*[src/controller/app_controller_2.js, line 164 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koiv5j5r20x-iI9ue7o:-Koiv5j6m3zWcwjaNjvQ:bvsx8ea) ([raw file](https://github.com/google/blockly-devtools/blob/06d2c885981fdf8ba774727d98bc332b35cceb6b/src/controller/app_controller_2.js#L164)):*
> ```JavaScript
>     const exporterTab = this.tabMap[AppController2.EXPORTER];
>     const workspaceFactoryTab = this.tabMap[AppController2.WORKSPACE_FACTORY];
> ```

Cache `this.view.selectedTab` in a local variable (`selectedTab`?) to reduce the visual clutter and reduce lookups.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/96)*
<!-- Sent from Reviewable.io -->

chynu committed:
Fixes for #96

chynu commented:
Review status: 9 of 14 files reviewed at latest revision, 9 unresolved discussions.

---

*[src/controller/app_controller_2.js, line 183 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koi7dvvZ67z2oOSkymS:-KoxfQWC5hV7Z_qQdeWg:bbdu1bw) ([raw file](https://github.com/google/blockly-devtools/blob/2beb5933f74fedb5f9968fcb9626fcf495d1481f/src/controller/app_controller_2.js#L183)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Sweet.  This got a lot smaller.

What happens if multiple tabs are told to show at the same time?
</blockquote></details>

Not sure when that might happen, considering one can only click on one tab at a time. It would just change views in sequence. (Unless I am misunderstanding your question?)

---

*[src/controller/app_controller_2.js, line 164 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/96#-Koiv5j5r20x-iI9ue7o:-KoxfOQs3SOxJxmNbZF1:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/06d2c885981fdf8ba774727d98bc332b35cceb6b/src/controller/app_controller_2.js#L164)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Cache `this.view.selectedTab` in a local variable (`selectedTab`?) to reduce the visual clutter and reduce lookups.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/96)*
<!-- Sent from Reviewable.io -->

rachel-fenichel commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 9 of 14 files reviewed at latest revision, 9 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/96#-:-KoxhIUgdLh0Q7zdwKwb:bnfp4nl)*
<!-- Sent from Reviewable.io -->
